libc/stdio_scanf.c

```diff
@@ -0,0 +1,4371 @@
+}
+
+
+TEST(stdio_scanf_squareBrackets, ascii)
+{
+	char buff[BUFF_LEN_STR] = { 0 };
+	char format[BUFF_LEN] = "%[^\n]%*c";
+	char res[BUFF_LEN_STR];
+	int i;
+
+	for (i = 1; i < 127; i++) {
+		if (i == 10) {
+			buff[i - 1] = i - 1;
+		}
+		else {
+			buff[i - 1] = i;
+		}
+	}
+
+	buff[i] = '\0';
+
+	fprintf(filep, "%s", buff);
+	rewind(filep);
+
+	/*
+	 * Scanf will try to find an occurrence of [] content and store it in res
+	 * After that it will discard all chars after [] content
+	 */
+	TEST_ASSERT_EQUAL_INT(1, test_vfscanfWrapper(filep, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(1, fscanf(filep, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+
+	TEST_ASSERT_EQUAL_INT(1, test_vsscanfWrapper(buff, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+
+	TEST_ASSERT_EQUAL_INT(1, sscanf(buff, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+
+
+	fclose(filep);
+	filep = fopen(TESTFILE_PATH, "w+");
+
+
+	memset(format, 0, sizeof(format));
+	memset(buff, 0, sizeof(buff));
+	memset(res, 0, sizeof(res));
+	for (i = 1; i < 127; i++) {
+		/* Bypass "]" because this sign end brackets formatension */
+		if (i == 93) {
+			buff[i - 1] = i - 1;
+		}
+
+		/* Bypass \n bcause of issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/679 */
+#ifdef __phoenix__
+		else if (i == 10) {
+			buff[i - 1] = i - 1;
+		}
+#endif
+
+		else {
+			buff[i - 1] = i;
+		}
+	}
+
+	fprintf(filep, "%s", buff);
+	rewind(filep);
+
+	sprintf(format, "%%[%s]", buff);
+
+	/*
+	 * Scanf will try to find all elements of ascii table in filep
+	 * (Without ] sign because it equals end of format) and store it in res
+	 */
+	TEST_ASSERT_EQUAL_INT(1, test_vfscanfWrapper(filep, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(1, fscanf(filep, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+
+	TEST_ASSERT_EQUAL_INT(1, test_vsscanfWrapper(buff, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+
+	TEST_ASSERT_EQUAL_INT(1, sscanf(buff, format, res));
+	TEST_ASSERT_EQUAL_STRING(buff, res);
+}
+
+
+TEST(stdio_scanf_squareBrackets, ranges)
+{
+	const char *buff = "loremIPSUM IPSUMdolor dolorSitAmet";
+	const char *buff2 = "123loremIPSUM IPSUMdolor123 dolor123SitAmet";
+	char res1[MAX_TESTSTR_WORDLEN], res2[MAX_TESTSTR_WORDLEN], res3[MAX_TESTSTR_WORDLEN];
+	char *format = "%[A-z] %[A-Z] %[a-z]";
+
+	fprintf(filep, "%s", buff);
+	rewind(filep);
+
+	/*
+	 * In this situation scanf will search all elements in range declared
+	 * inside brackets and stop right after it will be out of range.
+	 * (after finding a white sign in this case)
+	 */
+	TEST_ASSERT_EQUAL_INT(3, fscanf(filep, format, res1, res2, res3));
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM", res1);
+	TEST_ASSERT_EQUAL_STRING("IPSUM", res2);
+	TEST_ASSERT_EQUAL_STRING("dolor", res3);
+
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(3, test_vfscanfWrapper(filep, format, res1, res2, res3));
```

Comment:
for test to be more thorough you should reset the value of all output params before each call - right now you're assuming that if *scanf returns N it means N output params were set. If there would be an error of not setting one of them - the test would still pass because the variable holds the (valid) result) from the previous call.

This is cumbersome so I'm not enforcing the change right now - just be sure to know (and maybe write in comment) the current limitation.

Applies everywhere where *scanf function is called.

Reply:
A bit of work but added. Thanks 4 suggestion.